### ISSUE TITLE
Updating the source table when a route is advertised is racy

### ISSUE THREAD

**Issue by LeeSmet** [threefoldtech/mycelium#170]
Before sending an update to a peer, the source table needs to be updated to reflect the advertised route. This is needed to prevent routing loops, as subsequent received updates are checked against the source table and must be strictly better to be accepted (thus avoiding a loop since a route update from a loop can't contain a better route to begin with). Due to the concurrency primitives used, there are some slight alterations. Notably, the operation that needs to be performed on the source table is calculated, then the update is sent to the peer, and when all operations (if multiple updates are sent) have been calculated (and transmission of all of them has been queued), do we return and apply the required operations. In general this is fine, however in very low latency systems, it could be possible that the update is sent, received by the peer, and the peer sends us an update for this route, which we receive and process, before the source table could be updated. This is especially the case if a node adds itself as peer. If the original route for the subnet then degrades slightly (to the point where it is still accepted by the source table), it might become worse than the route advertised by the peer, effectively creating a loop.

In short, we need to make sure the source table is updated _before_ the update is actually sent. This might require some rearchitecting of the concurrency primitives used in the router.

**Comment by LeeSmet** [comment]
It would appear that the current public node setup with 2 nodes per location might also be affected by this
